### DISCUSSION TITLE
using-unix-domain-sockets-with-aspnetcore-and-httpclient/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#411] [Announcements]
# using-unix-domain-sockets-with-aspnetcore-and-httpclient/

In this post I show how to create an ASP.NET Core server that listens using Unix domain sockets, and how to call the server using an HttpClient

https://andrewlock.net/using-unix-domain-sockets-with-aspnetcore-and-httpclient/

**Comment by nmosafi** [comment]
Also worth noting that httpd can reverse proxy into a Unix domain socket using ProxyPass, and other reverse proxies support this. This is way cleaner than proxying onto a random port you happen to be listening to on the machine

**Comment by ChaosEngine** [comment]
This is nice and it works for me with Apache reverse proxy, however I could not make it work with SignalR websocket (upgrade socket, rewrite etc). So for simple req-resp this is great!

**Comment by JoaoCampos07** [comment]
Great article, probably this would be a good solution to use when you have your main application and them a worker process ?

**Comment by markusschaber** [comment]
Does anyone know whether browser, especially WebView2, supports this? Could be useful to have local apps with WebView based UIs more efficient and more secure (no port on localhost).

**Comment by fuyfuy** [comment]
"TCP/IP sockets are a limited resource; you can only have a maximum of 65,535 sockets in use at any time. When you add in the TIME_WAIT issue, the actual maximum number of active connections could be considerably smaller than this."
This is quite misleading - a total number of active TCP/IP connections can easily be in millions. What is limited to 65k is number of ports on which you can listen, but as long as connections arrive from different clients, you're not limited by 65k in any way.

**Reply by markusschaber** [reply]
On the other hand, you also need a port for the outgoing connection. So each client of a "localhost" connection eats up another port.
Carefully programmed clients might use the same port for several outgoing connections, but most programmers just don't spend the extra effort, and it only works to different servers - when connecting to the same server twice (e. G. http and websocket), the client needs different ports, because the quadruple (clientip, clientport, serverip, serverport) must be unique for each connection.

**Comment by sebgod** [comment]
Is it possible to use "normal" HTTP over TCP and HTTP over Unix domains sockets in the same kestrel instance?

**Comment by MithunChopda** [comment]
For production use, Is it okay to create and destroy a socket for every request or is it possible to pool sockets?

**Comment by MattParkerDev** [comment]
For anyone trying to get this to work in 2025, you'll need to set the HttpClient Version to http2:
```csharp
// Create an HttpClient using the UDS handler
var httpClient = new HttpClient(httpHandler)
{
	BaseAddress = new Uri("http://localhost"),
	DefaultRequestVersion = new Version(2, 0),
	DefaultVersionPolicy = HttpVersionPolicy.RequestVersionExact
};
```
